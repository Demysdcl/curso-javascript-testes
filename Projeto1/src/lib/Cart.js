import Dinero from 'dinero.js';
import find from 'lodash/find';
import remove from 'lodash/remove';

const Money = Dinero;

Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

const calculatePercentageDiscount = (quantity, condition, amount) => {
  if (quantity > condition.above) {
    return amount.percentage(condition.percentage);
  }
  return Money({ amount: 0 });
};

const calculateQuantityDiscount = (quantity, condition, amount) => {
  if (quantity > condition.quantity) {
    const isEven = quantity % 2 === 0;
    return amount.percentage(isEven ? 50 : 40);
  }
  return Money({ amount: 0 });
};

const calculateDiscount = (quantity, conditions, amount) => {
  const list = Array.isArray(conditions) ? conditions : [conditions];

  const [a] = list
    .map(c => {
      if (c.percentage) {
        return calculatePercentageDiscount(quantity, c, amount).getAmount();
      }
      if (c.quantity) {
        return calculateQuantityDiscount(quantity, c, amount).getAmount();
      }
    })
    .sort((a, b) => b - a);
  return Money({ amount: a });
};

export default class Cart {
  items = [];

  getTotal() {
    const total = this.items.reduce(
      (acc, { product: { price }, quantity, condition }) => {
        const amount = Money({ amount: price * quantity });

        let discount = Money({ amount: 0 });
        if (condition) {
          discount = calculateDiscount(quantity, condition, amount);
        }

        return acc.add(amount).subtract(discount);
      },
      Money({ amount: 0 }),
    );

    return {
      amount: total.toUnit(),
      formatted: total.toFormat('$0,0.00'),
    };
  }

  addProduct(item) {
    if (find(this.items, item)) {
      remove(this.items, item);
    }

    this.items.push(item);
  }

  removeProduct(item) {
    remove(this.items, ({ product }) => item === product);
  }

  checkout() {
    return {
      totalAmount: this.getTotal(),
      items: this.items,
    };
  }
}
