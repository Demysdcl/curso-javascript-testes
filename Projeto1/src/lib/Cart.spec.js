import Cart from './Cart';

describe('Shopping cart', () => {
  let cart;
  let product1;
  let product2;

  beforeEach(() => {
    cart = new Cart();

    product1 = {
      title: 'Adidas running shoes men',
      price: 35378,
    };

    product2 = {
      title: 'Adidas running shoes women',
      price: 42236,
    };
  });

  describe('getTotal', () => {
    it('should return 0 when cart is first created', () => {
      expect(cart.getTotal().amount).toEqual(0);
    });

    it('should return product price multiplied by product quantity', () => {
      cart.addProduct({ product: product1, quantity: 2 });
      expect(cart.getTotal().amount).toEqual(707.56);
    });

    it('should return a formatted amount for total', () => {
      cart.addProduct({ product: product1, quantity: 2 });
      cart.addProduct({ product: product1, quantity: 2 });
      expect(cart.getTotal().formatted).toEqual('R$707.56');
    });

    it('should update total when a product is included and then removed', () => {
      cart.addProduct({ product: product1, quantity: 2 });
      cart.removeProduct(product1);
      cart.addProduct({ product: product1, quantity: 1 });
      expect(cart.getTotal().amount).toEqual(353.78);
    });
  });

  describe('checkout', () => {
    it('should return a summary of all products, their quantities and the total amount', () => {
      cart.addProduct({ product: product1, quantity: 2 });
      cart.addProduct({ product: product2, quantity: 10 });
      expect(cart.checkout()).toMatchSnapshot();
    });
  });

  describe('special conditions', () => {
    it('should apply percentage discount', () => {
      const condition = {
        percentage: 30,
        above: 2,
      };

      const item = { product: product1, quantity: 3, condition };

      cart.addProduct(item);

      expect(cart.getTotal().amount).toEqual(742.94);
    });

    it('should not apply discount if condition is not met: percentage', () => {
      const condition = {
        percentage: 30,
        above: 2,
      };

      const item = { product: product1, quantity: 1, condition };

      cart.addProduct(item);

      expect(cart.getTotal().amount).toEqual(353.78);
    });

    it('should apply quantity discount for even quantity', () => {
      const condition = {
        quantity: 2,
      };

      const item = { product: product1, quantity: 4, condition };

      cart.addProduct(item);

      expect(cart.getTotal().amount).toEqual(707.56);
    });

    it('should apply quantity discount for odd quantity', () => {
      const condition = {
        quantity: 2,
      };

      const item = { product: product1, quantity: 5, condition };

      cart.addProduct(item);

      expect(cart.getTotal().amount).toEqual(1061.34);
    });

    it('should not apply discount if condition is not met: quantity', () => {
      const condition = {
        quantity: 2,
      };

      const item = { product: product1, quantity: 1, condition };

      cart.addProduct(item);

      expect(cart.getTotal().amount).toEqual(353.78);
    });

    it('should work as well with a mixed case', () => {
      let condition = {
        quantity: 2,
      };

      cart.addProduct({ product: product2, quantity: 9, condition });

      condition = {
        percentage: 30,
        above: 2,
      };

      cart.addProduct({ product: product1, quantity: 5, condition });

      expect(cart.getTotal().amount).toEqual(3518.97);
    });

    it('should apply the best condition when 2 or more conditions are supplied - first case', () => {
      const condition1 = {
        quantity: 2,
      };
      const condition2 = {
        percentage: 30,
        above: 2,
      };
      cart.addProduct({
        product: product1,
        quantity: 5,
        condition: [condition1, condition2],
      });

      expect(cart.getTotal().amount).toEqual(1061.34);
    });

    it('should apply the best condition when 2 or more conditions are supplied - second case', () => {
      const condition1 = {
        quantity: 2,
      };
      const condition2 = {
        percentage: 80,
        above: 2,
      };
      cart.addProduct({
        product: product1,
        quantity: 5,
        condition: [condition1, condition2],
      });

      expect(cart.getTotal().amount).toEqual(353.78);
    });
  });
});
