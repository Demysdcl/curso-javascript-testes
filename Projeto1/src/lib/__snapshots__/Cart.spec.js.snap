// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping cart checkout should return a summary of all products, their quantities and the total amount 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35378,
        "title": "Adidas running shoes men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 42236,
        "title": "Adidas running shoes women",
      },
      "quantity": 10,
    },
  ],
  "totalAmount": Object {
    "amount": 4931.16,
    "formatted": "R$4,931.16",
  },
}
`;
